### ISSUE TITLE
add health route api

### ISSUE THREAD

**Issue by tzzed** [telefonicaid/fiware-orion#4119]
**Is your feature request related to a problem / use case? Please describe.**
It will be great to add a `/health` route api for orion broker.

**Describe the solution you'd like**
Get the status of the broker can be very usefull.

**Describe alternatives you've considered**
Instead of health route I use the `/version` to know it orion broker is started.

**Describe why you need this feature**
A simple case is to know if the orion broker is started.

**Additional information**
Add any other information, diagrams or screenshots about the feature request here.

**Do you have the intention to implement the solution**
- [X] Yes, I have the knowledge to implement this new feature.
- [X] Yes, but I will need help.
- [ ] No, I do not have the skills.

**Comment by fgalan** [comment]
It is an interesting idea!

Just to know what you have in mind... which information do you would include in the health response? Could you even draft an example JSON on how it would look like?

Thanks for the feedback!

**Comment by fgalan** [comment]
As minor side note:

Note that the use case you describe:

>  A simple case is to know if the orion broker is started.

Can be solved with `/version` operations, that not only check that Orion is started but also how long it has been (in the `uptime`) field ;)

**Comment by tzzed** [comment]
It depends. But we can include some informations that are in `docker inspect`. 
something like this can be a motivation.
```js
 "State": {
            "Name": "/fiware-orion",
            "RestartCount": 0,
            "Driver": "overlay2",
            "Platform": "linux",
            "Status": "running",
            "Running": true,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 28700,
            "ExitCode": 0,
            "Error": "",
            "StartedAt": "2022-05-05T13:11:55.7592099Z",
            "FinishedAt": "0001-01-01T00:00:00Z",
      

}
```

**Comment by Anjali-NEC** [comment]
Hi @fgalan sir,

I would like to work on this issue, please let me know if this is still required or not? and please suggest what would be the  response of the `/health` API. Thanks

**Comment by fgalan** [comment]
The issue is still valid but it would require to analize what to return as response for the `/health` request. Unless you have a proposal for that (please add a comment with it in this issue in that case) I'd suggest to pick another issue.

**Comment by tzzed** [comment]
Hi @fgalan what about mixing some informations from `/version` and enities + subscription.

```json
{
"entities": "nb_entities",
"subscriptions": "nb_subscriptions",
"orion": {
    "version": "3.8.0",
    "uptime": "3 d, 8 h, 10 m, 33 s",
    "release_date": "Thu Jan 12 12:06:24 UTC 2023",
    "machine": "x86_64",
    "doc": "https://fiware-orion.rtfd.io/en/3.8.0/",   
}

```

**Comment by ArqamFarooqui110719** [comment]
Hi @fgalan, I also think it's a very good idea to implement `/health` API for Orion.

The purpose of `/health` API for any application is `to show the running status of the application`. Keeping that point in mind, I'd suggest that, `/version` API should keep the data of "version of Orion and it's dependency" and `/health` API should return "current running status of Orion and compilation info". For example:

Version API:
```
curl localhost:1026/version
```
version API Response after `/health` API:
```
  "version" : "3.x",
  "git_hash" : "xvz",
  "release_date" : "Mon Jan 29 08:45:34 UTC 2024",
  "doc" : "https://fiware-orion.rtfd.io/en/3.x.0/",
  "libversions": { xyz }

```

health API:
```
curl localhost:1026/health
```
health API response:
```
  "status" : "healthy/running"
  "uptime" : "x days"
  "compile_time" : "x days"
  "compiled_by" : "x days"
  "compiled_in" : "x days"
  "machine" : "x86_64"
  "port" : "1026"
```

I think with the above approach, the purpose of both the APIs will be fulfilled and also we will be able to add `/health` API. 

In case, there will be any new and valid information need to be added into any of these APIs, we can add that later also, if required. and for now we can go with the above details.

What is your opinion on this, @fgalan?

**Comment by ArqamFarooqui110719** [comment]
Hi @fgalan 
Please provide your opinion on above comment?

**Comment by fgalan** [comment]
In my opinion, it's not a good idea to touch the existing `GET /version` API request to remove fields. Some existing applications may be using these fields for health-checking so removing them (to move to that new `/health`) could be a potential backward compatibility problem with little gain.

(Side note: adding new fields to `/version` it is not risky)

Taking that into account, it would be possible to implement a `GET /health` operation if we find it useful. Which fields would include that operation (not already include in `/version` and in the the [statistics](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/admin/statistics.md) or [metrics](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/admin/metrics_api.md) APIs)? Could you provide a proposal (draft JSON), maybe re-formuling your previous proposal taking into account the requirement on not removing fields in the `GET /version` operation ?

**Comment by fgalan** [comment]
Taking into account the proposed fields:

* `compile_*` and `machine`: I think the machine and the time in which the ContextBroker binary was build is not relevant from the point of the view of the health of the service
* `port`: if the client is already getting this information as response of the `GET <host>:<port>/health` he already knows the port :)

Regarding the other ones:

* `status`: apart from `healthy` status, which other possibilities you suggest?
* `uptime`: fine with this.

Anyway, I think that only these two fields is a bit poor and doesn't justify the cost of developing a new operation in the API. I have the feeling that more information should be include here, providing information of the internal status of the contextBroker process, although I'm afraid I cannot provide more precise information (maybe you could take inspiration on some existing service with a similar health API?).

I think we need more suggestions on this before starting to implement this issue.

**Comment by ArqamFarooqui110719** [comment]
Will discuss on this once we will have more information. Meanwhile I'll look into some similar health APIs.

**Comment by fgalan** [comment]
Some ideas: consumed memory, used file descriptors, number of threads, etc.

**Comment by fgalan** [comment]
> Not sure about `database connectivity` field in `/health` API?

It could make sense.

**Comment by ArqamFarooqui110719** [comment]
> Some ideas: consumed memory, used file descriptors, number of threads, etc.

Yes, I was also thinking about these fields.

**Comment by fgalan** [comment]
Some additional ideas

* HealthCheck
* Connection status
* Alerts
* Service times
* Cardinalities (#N entities, #N subscriptions, #N services, etc.)
…

With regards to entities/subs in cardinality, I'm not sure. The per service/servicepath breakdown probably doesn't have sense for a health check (that breakdown makes more sense in metrics API), but the overall sum could be included.
